src/v/storage/batch_cache.h

```diff
@@ -34,6 +35,53 @@
 class batch_cache_test_fixture;
 namespace storage {
 
+namespace detail {
+
+class batch_cache_dirty_tracker {
+public:
+    void mark_dirty(std::pair<model::offset, model::offset> range) {
+        if (_min == model::offset{}) {
+            _min = range.first;
+            _max = range.second;
+        } else {
+            vassert(
+              _min < range.first && _max < range.second,
+              "newly tracked offset must be above any previously seen offsets "
+              "(inserting: [{}, {}], state: {})",
+              range.first,
+              range.second,
+              *this);
+            _max = range.second;
+        }
+    }
+
+    void mark_clean(model::offset up_to) {
+        if (up_to >= _max) {
+            _min = model::offset{};
+            _max = model::offset{};
+        } else {
+            _min = std::max(_min, up_to);
+        }
+    }
+
+    auto min() const { return _min; };
+    auto max() const { return _max; };
```

Comment:
The `max()` interface doesn't seem to be called in the PR, so I'm not sure how it is used exactly.

Is `max` basically the max last_offset of the batches in a range? If so, then maybe it would make sense to fold the entire dirty tracker logic and state into the range itself? It's minor, though I guess it's fine to have it as a separate structure.

Reply:
This is assuming we get rid of the tracker from the batch_cache_index and do a reverse iteration from `up_to` until we hit a clean range?

Reply:
Remove the dirty batch tracker usage from the batch_cache::range. Please see the updated cover letter in the commit https://github.com/redpanda-data/redpanda/pull/16954/commits/0409642204c74a975a0c429516267af6610b5194 which explains why it is kept for the index. Second to last section.